game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java

```diff
@@ -0,0 +1,141 @@
+package games.strategy.triplea.delegate.battle.steps.fire;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.delegate.Matches;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import lombok.Value;
+import org.triplea.java.collections.CollectionUtils;
+
+/**
+ * Group of units that are firing on targets
+ *
+ * <p>If the group is suicideOnHit, then all of the units must have the same unit type. This is
+ * because suicideOnHit units need to die when they hit a unit. If a firing group has multiple unit
+ * types, then it is harder to determine which unit type hit and so must commit suicide. By forcing
+ * the firing group to have all of the same unit type, it the units to commit suicide can be easily
+ * figured out since all of the units are the same.
+ */
+@Value
+public class FiringGroup {
+
+  @Value(staticConstructor = "of")
+  private static class SuicideAndNonSuicide {
+    Multimap<UnitType, Unit> suicideGroups;
+    Collection<Unit> nonSuicideGroup;
+
+    Collection<Collection<Unit>> values() {
+      final Collection<Collection<Unit>> values = new ArrayList<>(suicideGroups.asMap().values());
+      values.add(nonSuicideGroup);
+      return values;
+    }
+
+    int groupCount() {
+      return suicideGroups.keySet().size() + (nonSuicideGroup.isEmpty() ? 0 : 1);
+    }
+  }
+
+  String displayName;
+  String groupName;
+  Collection<Unit> firingUnits;
+  Collection<Unit> targetUnits;
+  boolean suicideOnHit;
```

Comment:
Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6278561)